src/Development/Rattle/Server.hs

```diff
@@ -194,12 +203,28 @@ cmdRattleStarted rattle@Rattle{..} cmd s msgs = do
             return (Right s, runSpeculate rattle >> go >> runSpeculate rattle)
 
 
+-- first is true; 2nd is false
+helper :: Monad m => (a -> m (Maybe b)) -> [a] -> m ([a], [b])
+helper _ [] = return ([], [])
+helper f (x:xs) = do
+  res <- f x
+  (as,bs) <- helper f xs
+  case res of
+    Nothing -> return (x:as, bs)
+    Just y -> return (as, y:bs)
+
+g :: (Hashable a, Eq a, Monad m) => (a -> m Bool) -> [a] -> m (Maybe (Set.HashSet a))
```

Comment:
Is this a filterM with a conditional on the outside, so [] => Nothing, xs => Just . Set.fromList?